django/db/migrations/autodetector.py

```diff
@@ -1153,6 +1157,25 @@ def _generate_added_field(self, app_label, model_name, field_name):
             and callable(field.default)
         ):
             self.questioner.ask_unique_callable_default_addition(field_name, model_name)
+
+        if field.generated:
+            expressions = list(field.expression.flatten())
+            f_references = {
+                expr.name for expr in expressions if isinstance(expr, models.F)
+            }
+            q_references = {
+                child[0].split(models.constants.LOOKUP_SEP, 1)[0]
+                for expr in expressions
+                if isinstance(expr, models.Q)
+                for child in expr.children
+                if isinstance(child, tuple)
+            }
```

Comment:
Nice, this looks to use the [approach](https://github.com/django/django/blob/7029b17315d74ea4a9e253f43485b8d73cd91067/django/db/models/query_utils.py#L179) from https://github.com/django/django/pull/16054/ 👍 

Maybe...
- @shangxiao you can pull out `referenced_base_fields` and it's tests as a separate commit
- I can merge that to main first
- You can both rebase your PRs and use it
- When I backport this, I would backport the commit for `referenced_base_fields` as well

Reply:
ok I can take a look 👍

Reply:
On reflection, I think this is not the best plan. Let's keep this as it is, and David when you rebase with these changes later on you can refact this to use `referenced_base_fields`

Reply:
Now in PR #18093 👍

Reply:
I'm very good at questioning myself. Wasn't sure it is necessary to backport two commits and a (slightly) larger change set 🤔 do you agree with the original approach?

Reply:
The original approach makes sense to me because the code that is being added in this PR is the exact same logic as the one factored out by david, and while there are surrounding tests, there are not specific tests for it. I don't see an issue in backporting the two commits as long as the one from David `Refs` this release blocker.